Source/DafnyLanguageServer/Workspace/SymbolTable.cs

```diff
@@ -123,21 +128,26 @@ public IEnumerable<Location> GetReferences(Uri uri, Position position) {
   }
 
   internal static Location? NodeToLocation(IHasNavigationToken node) {
-    if (node.NavigationToken.Uri == null) {
+    var origin = node.NavigationToken;
+    if (origin.Uri == null) {
       return null;
     }
     return new Location {
-      Uri = DocumentUri.From(node.NavigationToken.Uri),
-      Range = node.NavigationToken.GetLspRange()
+      Uri = DocumentUri.From(origin.Uri),
+      Range = origin.GetLspRange()
     };
   }
 
   public IHasNavigationToken? GetDeclarationNode(Uri uri, Position position) {
-    if (!nodePositions.TryGetValue(uri, out var forFile)) {
+    if (!navigationRanges.TryGetValue(uri, out var forFile)) {
       return null;
     }
+
+    var character = new Range(position, new Position(position.Line, position.Character + 1));
     return forFile.Query(position)
-      .Select(node => node is ISymbol ? node : ReferenceToNode.GetValueOrDefault(node))
+      .Where(n => n.ToLspRange().Intersects(character))
+      .Select(node => Declarations.TryGetValue(node, out var declaration)
+        ? declaration : ReferenceToNode.GetValueOrDefault(node))
```

Comment:
Since this code assumes there could be many references at the same position, would it make sense to sort them by smaller length before taking the first one?

Reply:
Generally this many references at the same location occurs because of AST nodes being copied and not being properly wrapped so they don't show up as copied, so the involved ranges are exactly the same.